README.md
# GoogleAuthenticator

[![Build Status](https://travis-ci.org/hanskorg/google-authenticator-rust.svg?branch=master)](https://travis-ci.org/hanskorg/google-authenticator-rust)
![Build Status](https://img.shields.io/crates/v/google-authenticator.svg)

## Introduction

This Rust crate can be used to interact with the Google Authenticator mobile app for 2-factor-authentication.This Rust crates can generate secrets, generate codes, validate codes and present a QR-Code for scanning the secret.It implements TOTP according to RFC6238
More about Google GoogleAuthenticator see:[Wiki](https://en.wikipedia.org/wiki/Google_Authenticator)


## Usage

Add this to your `Cargo.toml`:

```toml
[dependencies]
google-authenticator = "0.2.0"
```

## Examples

```rust
use google_authenticator::GoogleAuthenticator;

fn main() {
    let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
    let auth = GoogleAuthenticator::new();
    // let secret = auth.create_secret(32);
    let code = auth.get_code(&secret, 0).unwrap();

    assert!(auth.verify_code(&secret, &code, 1, 0).unwrap());
}
```

```rust
#[macro_use]
extern crate google_authenticator;
use google_authenticator::GA_AUTH;

fn main() {
    let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
    if let Ok(code) = get_code!(&secret) {
        println!("{}", verify_code!(&secret, &code, 1, 0));
    }
}
```

## Get the secret QR code

### Get Google Charts Url to make QR Code

```rust
use google_authenticator::{GoogleAuthenticator, ErrorCorrectionLevel};

fn main() {
    let auth = GoogleAuthenticator::new();
    let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
    println!(
        "{}",
        auth.qr_code_url(secret, "qr_code", "name", 200, 200, ErrorCorrectionLevel::High)
    );
}
```

```rust
#[macro_use]
extern crate google_authenticator;
use google_authenticator::GA_AUTH;

fn main() {
    let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
    println!("{}", qr_code_url!(&secret, "qr_code", "name"));
}
```

### Get QR code image in svg format

Change `Cargo.toml` to

```toml
[dependencies.google-authenticator]
version = "0.1.9"
features = ["with-qrcode"]
```

```rust
use google_authenticator::{GoogleAuthenticator, ErrorCorrectionLevel};

fn main() {
    let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
    let auth = GoogleAuthenticator::new();

    println!(
        "{}",
        auth.qr_code(secret, "qr_code", "name", 200, 200, ErrorCorrectionLevel::High)
            .unwrap()
    );
}
```

```rust
#[macro_use]
extern crate google_authenticator;
use google_authenticator::GA_AUTH;

fn main() {
    let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
    if let Ok(url) = qr_code!(&secret, "qr_code", "name") {
        println!("{}", url);
    }
}
```

## Contributors
Thanks to:
[JHZheng](https://github.com/zjhmale)  [Conbas](https://github.com/jtr109)

## FAQ
> You can post a new issue for help.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
This application follows the [Semantic Versioning standard](https://semver.org/).

## Unreleased
- 

## Version 0.3.0 (2021-11-15)
- Removed all unsafe code. (#9)
- Change in policy, there is no need for unsafe code in the future. (#9)
- `qrcode::EcLevel` implements `From<ErrorCorrectionLevel>` instead for
`ErrorCorrectionLevel` implementing `Into<qrcode::EcLevel>`
(only when "with-qrcode" feature is enabled). (#9)

## Pre version 0.3.0 (2020-10-12)
All changes before 2020-10-12 where not documented.
This is everything before and including: 388f6933301df132aaf94982b7f15c9dce3f2e06

src/authenticator.rs
// MIT License
//
// Copyright (c) 2018 hanskorg
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

use hmacsha1::hmac_sha1;
use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
use std::time::{SystemTime, UNIX_EPOCH};
use std::{error, fmt, result};

#[cfg(feature = "with-qrcode")]
use qrcode::render::svg;
#[cfg(feature = "with-qrcode")]
use qrcode::{EcLevel, QrCode};

#[cfg(any(feature = "with-qrcode"))]
use qrcode::types::QrError;

const SECRET_MAX_LEN: usize = 128;
const SECRET_MIN_LEN: usize = 16;

/// Controls the amount of fault tolerance that the QR code should accept. Require the feature
/// flag `with-qrcode`.
// This is a new enum to use in our public interface instead of rqcode::EcLevel.
#[derive(Copy, Clone)]
pub enum ErrorCorrectionLevel {
    /// 7% of data bytes can be restored.
    Low,
    /// 15% of data bytes can be restored.
    Medium,
    /// 25% of data bytes can be restored.
    Quartile,
    /// 30% of data bytes can be restored.
    High,
}

use self::ErrorCorrectionLevel::*;

impl fmt::Display for ErrorCorrectionLevel {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let result = match self {
            Low => 'L',
            Medium => 'M',
            Quartile => 'Q',
            High => 'H',
        };
        write!(f, "{}", result)
    }
}

#[cfg(feature = "with-qrcode")]
impl From<ErrorCorrectionLevel> for qrcode::EcLevel {
    fn from(level: ErrorCorrectionLevel) -> Self {
        match level {
            ErrorCorrectionLevel::High => EcLevel::H,
            ErrorCorrectionLevel::Medium => EcLevel::M,
            ErrorCorrectionLevel::Quartile => EcLevel::Q,
            ErrorCorrectionLevel::Low => EcLevel::L,
        }
    }
}

/// A list of all usable characters in base32.
const ALPHABET: [char; 33] = [
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
    'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '2', '3', '4', '5', '6', '7', '=',
];

/// The main interface of this library. It exports several function that are necessary to interface
/// with google authenticator.
pub struct GoogleAuthenticator {
    code_len: usize,
}

impl Default for GoogleAuthenticator {
    fn default() -> Self {
        Self { code_len: 6 }
    }
}

impl GoogleAuthenticator {
    /// Create a new `GoogleAuthenticator` using the default implementation. This means that the
    /// codes generated have a length of 6 and the secret will be chosen from allowed base32
    /// characters.
    ///
    /// ### Example
    /// ```
    /// use google_authenticator::GoogleAuthenticator;
    ///
    /// let auth = GoogleAuthenticator::new();
    /// ```
    pub fn new() -> Self {
        Self::default()
    }

    /// Use this method to configure the length of the generated code.
    ///
    /// ### Example
    /// ```rust
    /// use google_authenticator::GoogleAuthenticator;
    ///
    /// let auth = GoogleAuthenticator::new()
    ///     .with_code_length(8);
    /// ```
    pub fn with_code_length(mut self, code_length: usize) -> Self {
        self.code_len = code_length;
        self
    }

    /// Create new secret.
    ///
    /// Example:
    /// ```rust
    /// use google_authenticator::GoogleAuthenticator;
    ///
    /// let google_authenticator = GoogleAuthenticator::new();
    /// google_authenticator.create_secret(32);
    /// ```
    pub fn create_secret(&self, length: u8) -> String {
        let mut secret = Vec::<char>::new();
        let mut index: usize;
        for _ in 0..length {
            index = (rand::random::<u8>() & 0x1F) as usize;
            secret.push(ALPHABET[index]);
        }
        secret.into_iter().collect()
    }

    /// Calculate the code, with given secret and point in time. The `secret` parameter is the
    /// secret configured for this user. The `times_slice` parameter is the unix timestamp divided
    /// by 30 at which the code should expire.
    ///
    /// ### Example
    /// ```rust
    /// use google_authenticator::GoogleAuthenticator;
    ///
    /// let authenticator = GoogleAuthenticator::new();
    /// authenticator.get_code("I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3", 1523610659 / 30).unwrap();
    /// ```
    pub fn get_code(&self, secret: &str, times_slice: u64) -> Result<String> {
        if secret.len() < SECRET_MIN_LEN || secret.len() > SECRET_MAX_LEN {
            return Err(GAError::Error(
                "bad secret length. must be less than 128 and more than 16, recommend 32",
            ));
        }

        let message = if times_slice == 0 {
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs()
                / 30
        } else {
            times_slice
        };
        let key = Self::base32_decode(secret)?;
        let msg_bytes = message.to_be_bytes();
        let hash = hmac_sha1(&key, &msg_bytes);
        let offset = hash[hash.len() - 1] & 0x0F;
        let mut truncated_hash: [u8; 4] = Default::default();
        truncated_hash.copy_from_slice(&hash[offset as usize..(offset + 4) as usize]);
        let mut code = i32::from_be_bytes(truncated_hash);
        code &= 0x7FFFFFFF;
        code %= 1_000_000;
        let mut code_str = code.to_string();
        for i in 0..(self.code_len - code_str.len()) {
            code_str.insert(i, '0');
        }
        Ok(code_str)
    }

    /// This function verifies that a provided code is correct. The parameter `secret` is used to
    /// verify the user. `code` is the code that will be verified. The parameter `discrepancy`
    /// indicates number of seconds ago that a code may be generated. `time_slice` is used to modify
    /// what the current time is, as a unix timestamp. If 0 is provided here, the current time will
    /// be used.
    ///
    /// ### Example
    /// ```rust
    /// use google_authenticator::GoogleAuthenticator;
    ///
    /// let authenticator = GoogleAuthenticator::new();
    /// authenticator.verify_code("I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3", "224124", 3, 1523610659 / 30);
    /// ```
    pub fn verify_code(&self, secret: &str, code: &str, discrepancy: u64, time_slice: u64) -> bool {
        if code.len() != self.code_len {
            return false;
        }
        let curr_time_slice = if time_slice == 0 {
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs()
                / 30
        } else {
            time_slice
        };
        let start_time = curr_time_slice.saturating_sub(discrepancy);
        let end_time = curr_time_slice.saturating_add(discrepancy + 1);
        for _time_slice in start_time..end_time {
            if let Ok(c) = self.get_code(secret, _time_slice) {
                if code == c {
                    return true;
                }
            }
        }
        false
    }

    /// Get QR-Code URL for image, from google charts. For the height and width, if a value of 0 is
    /// provided, the default of `200px` is used. Level is the amount of fault tolerance that the
    /// QR code should accept, see
    /// [this page](https://en.wikipedia.org/wiki/QR_code#Error_correction) for more information.
    ///
    /// ### Example
    /// ```rust
    /// use google_authenticator::{GoogleAuthenticator, ErrorCorrectionLevel};
    ///
    /// let authenticator = GoogleAuthenticator::new();
    /// authenticator.qr_code_url(
    ///     "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3",
    ///     "your company name",
    ///     "hello",
    ///     0,
    ///     0,
    ///     ErrorCorrectionLevel::Medium,
    /// );
    /// ```
    pub fn qr_code_url(
        &self,
        secret: &str,
        name: &str,
        title: &str,
        width: u32,
        height: u32,
        level: ErrorCorrectionLevel,
    ) -> String {
        let width = if width == 0 { 200 } else { width };
        let height = if height == 0 { 200 } else { height };

        // Scheme will be a url to the totp we will use. Since it is a query parameter of the final
        // url, it must be percent encoded. In turn, this means that `name` and `title` must be
        // percent encoded twice for the final url to work if they contain characters that are not
        // allowed in urls.
        let scheme = Self::create_scheme(name, secret, title);
        let scheme = utf8_percent_encode(&scheme, NON_ALPHANUMERIC);
        format!(
            "https://chart.googleapis.com/chart?chs={}x{}&chld={}|0&cht=qr&chl={}",
            width, height, level, scheme
        )
    }

    /// Creates an in-memory SVG file that can be used to perform 2fa with Google Authenticator. The
    /// `height` and `width` parameters are the minimun dimensions of the generated svg. When 0 is
    /// supplied here, these values default to `200px`.
    ///
    /// ### Example
    /// ```rust
    /// use google_authenticator::{GoogleAuthenticator, ErrorCorrectionLevel};
    ///
    /// let authenticator = GoogleAuthenticator::new();
    /// authenticator.qr_code(
    ///     "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3",
    ///     "your company name",
    ///     "hello",
    ///     0,
    ///     0,
    ///     ErrorCorrectionLevel::Medium,
    /// );
    /// ```
    #[cfg(feature = "with-qrcode")]
    pub fn qr_code(
        &self,
        secret: &str,
        name: &str,
        title: &str,
        width: u32,
        height: u32,
        level: ErrorCorrectionLevel,
    ) -> Result<String> {
        let width = if width == 0 { 200 } else { width };
        let height = if height == 0 { 200 } else { height };
        let scheme = Self::create_scheme(name, secret, title);
        let code = QrCode::with_error_correction_level(scheme.as_bytes(), level.into())?;
        Ok(code
            .render()
            .min_dimensions(width, height)
            .dark_color(svg::Color("#000000"))
            .light_color(svg::Color("#ffffff"))
            .build())
    }

    /// Creates a totp url.
    fn create_scheme(name: &str, secret: &str, title: &str) -> String {
        let name = utf8_percent_encode(name, NON_ALPHANUMERIC);
        let title = utf8_percent_encode(title, NON_ALPHANUMERIC);
        format!("otpauth://totp/{}?secret={}&issuer={}", name, secret, title)
    }

    fn base32_decode(secret: &str) -> Result<Vec<u8>> {
        match base32::decode(base32::Alphabet::RFC4648 { padding: true }, secret) {
            Some(_decode_str) => Ok(_decode_str),
            _ => Err(GAError::Error("secret must be base32 decodeable.")),
        }
    }
}

/// Represents any of the reasons why using 2fa with Google Authenenticator can fail.
#[derive(Debug)]
pub enum GAError {
    /// An error in the logic of the QR code. Contains a static string with the error message.
    Error(&'static str),
    /// An error related to the QR code. This variant is only available with the feature flag
    /// `with-qrcode`.
    #[cfg(any(feature = "with-qrcode"))]
    QrError(QrError),
}

impl error::Error for GAError {
    fn description(&self) -> &str {
        match *self {
            GAError::Error(description) => description,
            #[cfg(any(feature = "with-qrcode"))]
            GAError::QrError(ref _err) => "",
        }
    }

    fn cause(&self) -> Option<&dyn error::Error> {
        match *self {
            #[cfg(any(feature = "with-qrcode"))]
            GAError::QrError(ref _err) => None,
            GAError::Error(_) => None,
        }
    }
}

#[cfg(any(feature = "with-qrcode"))]
impl From<QrError> for GAError {
    fn from(err: QrError) -> GAError {
        GAError::QrError(err)
    }
}

impl fmt::Display for GAError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            GAError::Error(desc) => f.write_str(desc),
            #[cfg(any(feature = "with-qrcode"))]
            GAError::QrError(ref err) => fmt::Display::fmt(err, f),
        }
    }
}

/// A type alias that can be used for fallible functions with `google_authenticator`.
pub type Result<T> = result::Result<T, GAError>;


src/google_authenticator.rs
//MIT License
//
//Copyright (c) 2018 hanskorg
//
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:
//
//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.
//
//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.
//
//

use rand;
use base32;
use hmacsha1::hmac_sha1;
use std::mem;
use std::time::{SystemTime, UNIX_EPOCH};
use urlencoding;

#[cfg(any(feature = "with-qrcode"))]
use qrcode::{QrCode, Version, EcLevel};
#[cfg(any(feature = "with-qrcode"))]
use qrcode::render::svg;

const SECRET_MAX_LEN:usize = 128;
const SECRET_MIN_LEN:usize = 16;

pub struct GoogleAuthenticator{
    code_len:usize,
    _base32_alphabet:Vec<char>
}

impl GoogleAuthenticator{

    pub fn new() -> GoogleAuthenticator{
        GoogleAuthenticator{
            code_len: 6,
            _base32_alphabet: vec![
                'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', //  7
                'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', // 15
                'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', // 23
                'Y', 'Z', '2', '3', '4', '5', '6', '7', // 31
                '=',
            ]
        }
    }
    /// Create new secret.
    /// characters, randomly chosen from the allowed base32 characters.
    ///
    /// Example:
    ///```
    /// use google_authenticator::GoogleAuthenticator;
    ///
    /// let google_authenticator = GoogleAuthenticator::new();
    /// google_authenticator.create_secret(32);
    ///
    ///```
    ///
    pub fn create_secret(&self, length:u8) -> String{
        let mut secret = Vec::<char>::new();
        let mut index:usize ;
        for _ in 0 .. length {
            index = (rand::random::<u8>() & 0x1F) as usize;
            secret.push(self._base32_alphabet[ index ]);
        }
        secret.into_iter().collect()
    }

    /// Calculate the code, with given secret and point in time.
    ///
    /// Example:
    ///```
    ///     use google_authenticator::GoogleAuthenticator;
    ///
    ///     let google_authenticator = GoogleAuthenticator::new();
    ///     google_authenticator.get_code("I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3", 1523610659 / 30).unwrap();
    ///
    /// ```
    /// *secret* : user secret, it will verify each user.
    /// *times_slice* : unix_timestamp / 30 ,if give 0, it will system unix_timestamp
    ///
    pub fn get_code(&self, secret:&str, times_slice:i64) -> Result<String>{

        if secret.len() < SECRET_MIN_LEN || secret.len() > SECRET_MAX_LEN {
            return Err(GAError::Error("bad secret length. must be less than 128 and more than 16, recommand 32"));
        }

        let mut message: i64 = times_slice;
        if times_slice == 0 {
            message = (SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64) / 30;
        }
        // println!("{:?}", message);
        let key  = self._base32_decode(secret)?;
        let msg_bytes = message.to_be_bytes();//unsafe { mem::transmute::<i64,[u8;8]>(message.to_be()) };
        // let mut message_body:Vec<u8> = vec![0;4];
        // for msg_byte in msg_bytes.iter() {
        //     message_body.push(*msg_byte);
        // }
        let hash = hmac_sha1(&key, &msg_bytes);

        let offset = hash[hash.len() - 1] & 0x0F;
        let mut truncated_hash:[u8;4] = Default::default();
        truncated_hash.copy_from_slice(&hash[offset  as usize .. (offset + 4)  as usize]);
        let mut code:i32 = unsafe { mem::transmute::<[u8;4], i32>(truncated_hash) };
        if cfg!(target_endian = "big") {
        } else {
            code = i32::from_be(code);
        }
        code = code & 0x7FFFFFFF;
        code =  code % 1_000_000i32;
        let mut code_str = code.to_string();        
        for i in 0 .. (self.code_len - code_str.len()) {
            code_str.insert(i,'0');
        }
        Ok(code_str)
    }
    /// Check if the code is correct.
    /// `secret` use for verify user
    /// `code` the code to verify
    /// `discrepancy` This will accept codes starting from *discrepancy\*30sec* ago to *discrepancy\*30sec* from now.
    /// `time_slice` if give 0, it will system unix_timestamp
    ///
    ///```
    ///     use google_authenticator::GoogleAuthenticator;
    ///
    ///     let google_authenticator = GoogleAuthenticator::new();
    ///     google_authenticator.verify_code("I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3", "224124", 3, 1523610659 / 30);
    ///
    /// ```
    ///
    pub fn verify_code(&self, secret:&str, code: &str, discrepancy:i64, time_slice:i64) -> bool {
        if code.len() != self.code_len {
            return false;
        }
        let mut curr_time_slice: i64 = time_slice;
        if time_slice == 0 {
            curr_time_slice = (SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64) / 30 ;
        }
        for _time_slice in curr_time_slice.wrapping_sub(discrepancy) .. curr_time_slice.wrapping_add(discrepancy + 1)  {
            if let Ok(c) = self.get_code(secret, _time_slice) {
                if code.eq(c.as_str()) {
                    return true;
                }
            }
        }
        false
    }
    /// Get QR-Code URL for image, from google charts.
    /// width: width of the qrcode. default value 200 px
    /// height: height of the qrcode. default value 200 px
    /// level: the qrcode level ,it will be L,M,Q,H. Default value is M
    ///
    pub fn qr_code_url(&self, secret:&str, name:&str, title:&str, width:u16, height:u16, level:char) -> String {
        let _width = if width == 0 {200} else {width};
        let _height = if  height == 0 {200} else {height};
        let levels = vec!['L', 'M', 'Q', 'H'];
        let _level = if levels.contains(&level) {level} else {'M'};
        let scheme =  urlencoding::encode(
            format!("otpauth://totp/{}?secret={}&issuer={}"
                    ,name
                    ,secret
                    ,urlencoding::encode(title)).as_str());
        format!("https://chart.googleapis.com/chart?chs={}x{}&chld={}|0&cht=qr&chl={}", _width, _height, level, scheme)
    }
    /// Get QR-Code  for svg
    /// Get QR-Code URL for image, from google charts.
    /// width: width of the qrcode. default value 200 px
    /// height: height of the qrcode. default value 200 px
    /// level: the qrcode level ,it will be L,M,Q,H. Default value is M
    #[cfg(any(feature = "with-qrcode"))]
    pub fn qr_code(&self, secret:&str, name:&str, title:&str, width:u16, height:u16, level:char) -> Result<String>{
        let _width = if width == 0 {200} else {width};
        let _height = if  height == 0 {200} else {height};
        let levels = vec!['L', 'M', 'Q', 'H'];
        let _level = match level {
            'L' => EcLevel::L,
            'H' => EcLevel::H,
            _ => EcLevel::M
        };
        let scheme = format!("otpauth://totp/{}?secret={}&issuer={}", name, secret, title);
        let code = QrCode::with_error_correction_level(scheme.as_bytes(), _level)?;
        Ok(code.render()
            .min_dimensions(_width as u32, _height  as u32)
            .dark_color(svg::Color("#000000"))
            .light_color(svg::Color("#ffffff"))
            .build())
    }

    fn _base32_decode(&self, secret:&str) -> Result<Vec<u8>>{
        //use base32 extern
        match  base32::decode(base32::Alphabet::RFC4648 { padding: true }, secret) {
            Some(_decode_str) => Ok(_decode_str),
            _                 => Err(GAError::Error("secret must can decode by base32."))
        }
    }

}


use std::error;
use std::result;
use std::fmt;
#[cfg(any(feature = "with-qrcode"))]
use qrcode::types::QrError;
#[derive(Debug)]
pub enum GAError {
    Error(&'static str),
    #[cfg(any(feature = "with-qrcode"))]
    QrError(QrError)
}

impl error::Error for GAError{
    fn description(&self) -> &str {
        match *self {
            GAError::Error(description) =>description,
            #[cfg(any(feature = "with-qrcode"))]
            GAError::QrError(ref err) => ""
        }
    }

    fn cause(&self) -> Option<&dyn error::Error> {
        match *self {
            #[cfg(any(feature = "with-qrcode"))]
            GAError::QrError(ref err) =>  None,
            GAError::Error(_) => None,
        }
    }
}
#[cfg(any(feature = "with-qrcode"))]
impl From<QrError> for GAError{
    fn from(err:QrError) -> GAError {
        GAError::QrError(err)
    }
}

impl fmt::Display for GAError{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            GAError::Error(desc) => f.write_str(desc),
            #[cfg(any(feature = "with-qrcode"))]
            GAError::QrError(ref err) =>  fmt::Display::fmt(err, f),
        }
    }
}

type Result<T> = result::Result<T,GAError>;


src/lib.rs
#![deny(missing_docs)]
#![deny(unsafe_code)]
// Copyright 2012-2013 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! This Rust crate can be used to interact with the Google Authenticator mobile app for 2-factor-authentication.
//! This Rust crates can generate secrets, generate codes, validate codes and present a QR-Code for scanning the secret.
//! It implements TOTP according to RFC6238
//!
//! # Examples
//!
//! ```rust
//! use google_authenticator::GoogleAuthenticator;
//!
//! let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
//!
//! let auth = GoogleAuthenticator::new();
//! let code = auth.get_code(secret,0).unwrap();
//! if auth.verify_code(secret, code.as_str(), 1, 0) {
//!     println!("match!");
//! }
//! ```

mod authenticator;

pub use authenticator::*;

lazy_static::lazy_static! {
    /// A globally accessible, thread safe instance of a `GoogleAuthenticator`. Note that if the
    /// code panics while this variable is in scope, the `std::sync::Mutex` can be poisoned,
    /// preventing further access to this variable.
    pub static ref GA_AUTH: GoogleAuthenticator = GoogleAuthenticator::new();
}

/// A macro that can be used for convenient access to the function
/// `GoogleAuthenticator::create_secret`, by providing a default of `32` to the `length` parameter.
#[macro_export]
macro_rules! create_secret {
    ($length: expr) => {
        GA_AUTH.create_secret($length)
    };
    () => {
        GA_AUTH.create_secret(32)
    };
}

/// A macro that can be used for convenient access to the function
/// `GoogleAuthenticator::get_code`, by providing a default of the current time to the
/// `times_slice` parameter.
#[macro_export]
macro_rules! get_code {
    ($secret: expr, $time_slice: expr) => {
        GA_AUTH.get_code($secret, time_slice)
    };
    ($secret: expr) => {
        GA_AUTH.get_code($secret, 0)
    };
}

/// A macro that can be used for convenient access to the function
/// `GoogleAuthenticator::verify_code`, by providing a default of 0 to the `discrepancy` parameter,
/// and the current time to the `times_slice` parameter.
#[macro_export]
macro_rules! verify_code {
    ($secret: expr, $code: expr, $discrepancy: expr, $time_slice: expr) => {
        GA_AUTH.verify_code($secret, $code, $discrepancy, $time_slice)
    };
    ($secret: expr, $code: expr) => {
        GA_AUTH.verify_code($secret, $code, 0, 0)
    };
}

/// A macro that can be used for convenient access to the function
/// `GoogleAuthenticator::qr_code_url`, by providing a default of 200 to the `width` parameter, 200
/// to the `height` parameter, and `ErrorCorrectionLevel::Medium` to the `level` parameter.
#[macro_export]
macro_rules! qr_code_url {
    ($secret: expr, $name: expr, $title: expr, $width: expr, $height: expr, $level: expr) => {
        GA_AUTH.qr_code_url($secret, $name, $title, $width, $height, $level)
    };
    ($secret: expr, $name: expr, $title: expr) => {
        GA_AUTH.qr_code_url(
            $secret,
            $name,
            $title,
            200,
            200,
            crate::ErrorCorrectionLevel::Medium,
        )
    };
}

/// A macro that can be used for convenient access to the function
/// `GoogleAuthenticator::qr_code`, by providing a default of 200 to the `width` parameter, 200
/// to the `height` parameter, and `ErrorCorrectionLevel::Medium` to the `level` parameter.
#[macro_export]
macro_rules! qr_code {
    ($secret: expr, $name: expr, $title: expr, $width: expr, $height: expr, $level: expr) => {
        GA_AUTH.qr_code($secret, $name, $title, $width, $height, $level)
    };
    ($secret: expr, $name: expr, $title: expr) => {
        GA_AUTH.qr_code(
            $secret,
            $name,
            $title,
            200,
            200,
            crate::ErrorCorrectionLevel::Medium,
        )
    };
}

#[cfg(test)]
mod tests {
    #[cfg(feature = "with-qrcode")]
    use crate::ErrorCorrectionLevel::*;
    use crate::GoogleAuthenticator;

    #[test]
    fn create_secret() {
        let auth = GoogleAuthenticator::new();
        let secret = auth.create_secret(32);
        //auth.get_code(secret.as_str(),0);
        //        println!("{:?}",secret);
        assert_eq!(secret.len(), 32);
    }

    #[test]
    fn test_code() {
        let auth = GoogleAuthenticator::new();
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        assert_eq!(6, auth.get_code(secret, 0).unwrap().len());
    }

    #[test]
    fn test_verify_code() {
        let auth = GoogleAuthenticator::new();
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        #[cfg(feature = "with-qrcode")]
        println!(
            "{:?}",
            auth.qr_code(secret, "qr_code", "name", 0, 0, Medium)
        );
        assert!(auth.verify_code(secret, "224124", 3, 1523610659 / 30));
    }

    #[test]
    #[cfg(feature = "with-qrcode")]
    fn test_qr_code_url() {
        let auth = GoogleAuthenticator::new();
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        let url = auth.qr_code_url(secret, "secret code", "hi there", 0, 0, Medium);
        println!("{}", url);
        let resp = ureq::get(&url).call();
        assert!(resp.ok());
        // panic!();
    }

    #[test]
    #[cfg(feature = "with-qrcode")]
    fn test_qr_code() {
        let auth = GoogleAuthenticator::new();
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        auth.qr_code(secret, "secret_code", "hi", 0, 0, Medium)
            .unwrap();
    }
}

#[cfg(test)]
mod macro_tests {
    use crate::GA_AUTH;

    #[test]
    fn create_secret() {
        let secret = create_secret!();
        assert_eq!(secret.len(), 32);
    }

    #[test]
    fn test_code() {
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        assert_eq!(6, get_code!(secret).unwrap().len());
    }

    #[test]
    fn test_verify_code() {
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        let code = get_code!(secret).unwrap();
        assert!(verify_code!(secret, &code));
    }

    #[test]
    #[cfg(feature = "with-qrcode")]
    fn test_qr_code() {
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        assert!(qr_code!(secret, "qr_code", "name").is_ok());
    }

    #[test]
    #[cfg(feature = "with-qrcode")]
    fn test_qr_code_url() {
        let secret = "I3VFM3JKMNDJCDH5BMBEEQAW6KJ6NOE3";
        qr_code_url!(secret, "qr_code", "name");
    }
}
